resolwe/flow/managers/base.py

```diff
@@ -87,11 +198,133 @@ def _include_environment_variables(self, program):
 
-    def communicate(self, run_sync=False, verbosity=1):
+    def _get_per_data_dir(self, dir_base, data_id):
+        """Return the given base directory extended with a per-data component."""
+        result = getattr(settings, 'FLOW_EXECUTOR', {}).get(dir_base, '')
+        return os.path.join(result, str(data_id))
+
+    def _prepare_data_dir(self, data_id):
+        """Prepare destination directory where the data will live."""
+        output_path = self._get_per_data_dir('DATA_DIR', data_id)
+        dir_mode = getattr(settings, 'FLOW_EXECUTOR', {}).get('DATA_DIR_MODE', 0o755)
+        os.mkdir(output_path, mode=dir_mode)
+        # os.mkdir is not guaranteed to set the given mode
+        os.chmod(output_path, dir_mode)
+        return output_path
+
+    def _prepare_context(self, data_id, data_dir, runtime_dir, **kwargs):
+        """Prepare settings and constants jsons for the executor."""
+        settings_dict = dict(kwargs)
+        settings_dict.update(self._marshal_settings())
+        settings_dict['DATA_DIR'] = data_dir
+
+        # add scheduling classes
+        settings_dict['Process_sched_class'] = {
+            k: getattr(Process, k) for k in dir(Process)
+            if k.startswith('SCHEDULING_CLASS_') and isinstance(getattr(Process, k), str)
+        }
+        # add Data status constants
+        settings_dict['Data_status'] = {
+            k: getattr(Data, k) for k in dir(Data)
+            if k.startswith('STATUS_') and isinstance(getattr(Data, k), str)
+        }
+        # required plumbing settings
+        settings_dict['REDIS_CHANNEL_PAIR'] = state.MANAGER_EXECUTOR_CHANNELS
+
+        # extend the settings with whatever the executor wants
+        settings_dict.update(self.executor.make_settings(data_id))
+
+        # save the settings into a json in the data dir
+        json.dump(settings_dict, open(os.path.join(runtime_dir, 'settings.json'), 'wt'), cls=SettingsJSONifier)
+
+    def _prepare_executor(self, data_id, executor):
+        """Copy executor sources into the destination directory."""
+        import resolwe.flow.executors as executor_module
+
+        exec_dir = executor_module.__path__[0]
+        dest_dir = self._get_per_data_dir('RUNTIME_DIR', data_id)
+        module_dir = os.path.join(dest_dir, 'executors')
+        shutil.copytree(exec_dir, module_dir)
+        dir_mode = getattr(settings, 'FLOW_EXECUTOR', {}).get('RUNTIME_DIR_MODE', 0o755)
+        os.chmod(dest_dir, dir_mode)
+
+        class_name = executor.split('.')
+        class_name = class_name[class_name.index('executors') + 1:]
+        return '.' + '.'.join(class_name) + '.run.FlowExecutor', dest_dir
+
+    def _prepare_script(self, dest_dir, program):
+        """Copy the execution script for this data into the destination directory."""
+        script_name = 'task.script'
+        dest_file = os.path.join(dest_dir, script_name)
+        open(dest_file, 'wt').write(program)
```

Comment:
When will this file descriptor be closed?

Reply:
Immediately after this line, since the object's reference count immediately drops to 0.

Reply:
Well, the reference count does drop to 0, but it doesn't mean that the file is immediately garbage collected AFAIK?

Reply:
Since reference counting is an internal detail of CPython, it would be better to use a context manager.

Reply:
Garbage collection is an internal implementation detail anyway, whether in plain code or with context managers. There are no lifetime guarantees anywhere, other than being alive at least as long as the reference count is positive. As far as CPython is concerned, the file object will be destroyed immediately after that line, same as with a context manager.

Reply:
> ... or with context managers. There are no lifetime guarantees anywhere

This is true of the file object, but the point here is that there is an underlying file descriptor resource. In the case of a context manager, the file descriptor is closed after exiting the block, regardless of then the actual file object is garbage collected.